### ISSUE TITLE
make tergmLite use ergm terms

### ISSUE THREAD

**Issue by chad-klumb** [statnet/tergmLite#23]
It appears to me that tergmLite could be made to use ergm terms with relatively minor modifications to each.

Specifically, tergmLite will need to call the relevant InitErgmTerm function when updating term inputs (rather than the currently used updateInputs_termname function), and should allow default argument settings to be handled by check.ErgmTerm (rather than get_formula_term_args_in_formula_env, which currently sets defaults in tergmLite).

ergm will need to make these InitErgmTerm functions available (exporting them, or exporting a function that can call them), and will also need to accommodate ```networkLite``` representations of network data in the relevant portions of ergm.  The latter can be accomplished by replacing calls to ```network``` accessors (like %v%, %n%, is.directed, is.bipartite, network.size, network.dyadcount, etc.) with calls to generic functions that can dispatch to ```network``` methods or appropriate ```networkLite``` analogues.  These changes need to be made in check.ErgmTerm and the relevant InitErgmTerm functions.  Such changes should not affect ergm's behavior on ```network``` objects, but should open up the modified parts of ergm to other network data representations (including ```networkLite``` objects).

I have made changes along the above lines to tergmLite and ergm and will be pushing those to new branches (rather than forks, per request) shortly for scrutiny, testing, and discussion.  I have defined a (minimal) ```nds``` API (nds = network data structure or network data storage) which defines the generic accessors, and placed those definitions in ergm (for the time being), along with the ```network``` class methods.  The ```networkLite``` class methods are in tergmLite, which also houses a trivial ```networkLite``` constructor.

Loosely related are term updates in EpiModel (which will also be placed in their own branch).  I would recommend getting the updated (branch) versions of all three packages if testing.  Incidentally, if we agree to these changes, EpiModel and tergmLite will need their ergm version dependencies updated to whatever the version of ergm is after these changes are made.

For now tergmLite will only support the terms it already did, but those terms will use the corresponding InitErgmTerm function.  More ergm terms could be opened up to tergmLite relatively easily after these changes (just change the accessors in InitErgmTerm, and change the list of allowed terms in tergmLite).

I think these changes bring up many questions, like how closely ergm should be tied to network, and whether we should define a more complete template class or API for network data representations that can be plugged into ergm.  My guess is that some parts of ergm will be easier to make "generic" than others, but even if these changes don't go beyond the terms themselves, I still think it's worth the effort to make things more flexible.

**Comment by chad-klumb** [comment]
build errors appear related to ```nlme``` version

apparently a similar thing happened a few years ago (might have some ideas for fixing):

https://github.com/travis-ci/travis-ci/issues/5756

**Comment by krivit** [comment]
I've already put this in a commit comment, but just to make sure everyone gets the message, this is not a good idea, and someone should have checked with me before starting to implement this. We are already in the process of making the `network` package methods generic, so efforts should be directed there and not into the `nds_*` functions.

**Comment by krivit** [comment]
@CarterButts , @martinamorris , @smjenness

**Comment by chad-klumb** [comment]
Okay, I can try to make changes in network instead.

The nds thing was just my way of thinking about generalizing access to network information, which I think we do want to do.  I didn't know it was already underway in network.

We will still need to make InitErgmTerm functions available to tergmLite though.

**Comment by krivit** [comment]
Here's how I'd do it:

1. Make `network` functions generic.
2. Implementing and export the appropriate methods from `tergmLite`.
3. Make whatever tweaks to be able to call `ergm_model()` on `tergmLite`'s networks.

Right now, we have `pending_update_network` API in `ergm` as well.

**Comment by chad-klumb** [comment]
Update:

I have shifted generic definitions from ```ergm``` to ```network```.  By tweaking ```networkLite```'s data structure to more closely resemble ```network```'s, only two new generics were needed:

```list.vertex.attributes```
```get.vertex.attribute```

Those are now defined as generics within ```network```, with the existing implementations becoming the ```network``` class methods.  ```networkLite``` defines its own versions of these two methods, reflective of the different list structure used to store vertex attributes (```network```'s being the transpose of ```networkLite```'s).  

While only one more line of code would be needed to make ```networkLite```'s vertex attribute storage identical to ```network```'s (and thus compatible with existing ```network``` class vertex attribute functions), it would mean copying all the vertex attributes every time a ```networkLite``` object is constructed, which is slow (so I haven't done it that way).

In the end no changes were needed to ```ergm``` so I have deleted that branch.  To work with ```ergm_model``` it was necessary to implement ```as.network.networkLite``` (which has been added to ```tergmLite```).

By pursuing changes in ```tergmLite```'s ```updateModelTermInputs``` function a little farther than I already had, I have made it use ```ergm_model``` to access the ```InitErgmTerm``` functionality.  The assumption that the dissolution model is the initial segment of the formation model continues to be used in updating the terms.  The ```get_formula_term_args_in_formula_env``` function is no longer used (this being handled within ```ergm_model```), so that function (and its tests) could safely be deleted from the standpoint of ```tergmLite``` itself.  That function is, however, exported, so it could conceivably be being used in other people's code, so I've left it in ```tergmLite``` for the time being.

I do not currently have push access on ```network```.  If I could get that soon it would allow me to make a branch with the ```network``` changes (otherwise I'll have to fork).  I would like to do some more testing but can get all the changes up tomorrow.  There will be new versions of ```network``` and ```tergmLite``` that will both be needed for testing; term updates to ```EpiModel``` remain as before, and are not strictly necessary to test changes to ```tergmLite```.

**Comment by chad-klumb** [comment]
travis CI errors notwithstanding I think this is ready for testing; everything passed R CMD check on my computer

you will want:

chad-klumb/network@tergmLiteterms
statnet/tergmLite@tergmLiteterms

and optionally

statnet/EpiModel@tergmLiteterms

Be sure you have ERGM version 3.10.

A couple of remarks about EpiModel:

the term absdiffby had some logic in its updateInputs function (formerly in tergmLite) that was not matched in its InitErgmTerm function (in EpiModel).  Basically it had some limited functionality for converting a "by" argument that wasn't binary-valued into one that was.  Since this was only in updateInputs (and not InitErgmTerm) I'm not sure how much use it would have been getting, but if there's a need for that sort of functionality we can add it into the InitErgmTerm function with a levels argument that allows the user to specify which categories should be treated as the 1 category (or 0 category if that's preferred).  (Actually the changestat function only checks if something is equal to 1, so what you call the other category doesn't matter; -1 might be better than 0, given how the asymmetry is added in.)

other point is about initialize.net; there's a point where vertex attributes are copied over

https://github.com/statnet/EpiModel/blob/2c131f04390f7a2ceae19711e7fa43792f06e2e6/R/net.mod.init.R#L60-L63

A closer look at `get_formula_term_attr` shows that it greps the formula for attribute names:

https://github.com/statnet/EpiModel/blob/2c131f04390f7a2ceae19711e7fa43792f06e2e6/R/net.utils.R#L567-L585

However, both the old and new UI permit formulas that don't contain all the attribute names they depend on.  The old UI might have something like
```
x <- "myfavoriteattrname"
valid.formula <- ~ edges + concurrent(by = x)
```
The new UI might have something like
```
f <- function(x) log(1 + exp(x %v% "attrname")^2)
valid.formula <- ~ edges + nodecov(f)
```
Both of those work in ergm, but the necessary attributes aren't copied over by initialize.net, since their names don't explicitly appear in the formula.

There might be a slick way to expand the RHS of the formula and see all the dependent attribute names but I don't know how to do that, so I've changed the code in initialize.net to copy over all the vertex attributes except the three that were already explicitly excluded (namely ```c("active", "vertex.names", "na")```).  I don't think this will be a performance issue since it only happens in initialize.net (but let me know if I'm wrong).

**Comment by sgoodreau** [comment]
Chad - thanks much for this!  I am back in a position of being able to test it out, and indeed I might already need to be using it for research-level work in the coming days. To confirm the latter, can you clarify my understanding of one thing? Based on the following:

> There might be a slick way to expand the RHS of the formula and see all the dependent attribute names but I don't know how to do that, so I've changed the code in initialize.net to copy over all the vertex attributes except the three that were already explicitly excluded (namely `c("active", "vertex.names", "na")`). I don't think this will be a performance issue since it only happens in initialize.net (but let me know if I'm wrong).

I take it to mean that the branches/forks discussed in here (chad-klumb/network@tergmLiteterms
and statnet/tergmLite@tergmLiteterms) should allow one to use the new UI terms such as nodefactor(~age<25) in conjunction with tergmLite.  Is that right?

**Comment by chad-klumb** [comment]
Yes, they should allow one to use the new attr and levels UI for ergm terms that are allowed by tergmLite, as processing of term arguments is now handled by the corresponding InitErgmTerm function.

**Comment by sgoodreau** [comment]
On first pass, this seems to be working with my research-level runs, which include a nodefactor(~age>25) term in both the formation and dissolution models.  By "on first pass", I mean that prior to downloading these various branches/forms, I got an error about my attribute name not being in the named list, which I tracked down to coming from tergmLite, then I remembered this github thread, came here, saw that you'd done this, grabbed all the right versions, and now my code proceeds without error.  It is still underway (and will be for some time), so I can't check yet whether the networks as simulated behave as expected, but the fact that it gets over the previous error is encouraging.  Thanks!!

**Comment by smjenness** [comment]
@chad-klumb I got all the tests to pass locally with the branches set as above. I am trying now to update the .travis.yml to pull from the right EpiModel branch to pass TravisCI.

**Update: it now passes TravisCI**

**Comment by smjenness** [comment]
@chad-klumb One thing that needs to be checked, updated is that your modification to EpiModel now does not pass for that package. I think the modifications to work with the new API need to be merged into master in this order:

- [ ] EpiModel
- [ ] network
- [ ] tergmLite
- [ ] EpiModelHIV

**Comment by smjenness** [comment]
@chad-klumb I added a full-scale (research-level) workflow with EpiModelHIV under the `inst` subfolder (it's too big to add to the unit tests) on your branch. Please run and test the new changes there. One problem that I am seeing already, just at the estimation phase, is that the estimation of these large networks is much slower than the current (ergm 3.9, tergmLite master) approach. In particular, the fitting of the first (main partnership) model gets stuck here for at least a half-hour (after which point I give up and cancel it):

```
Fitting initial model.
Unable to match target stats. Using MCMLE estimation.
Starting maximum pseudolikelihood estimation (MPLE):
Evaluating the predictor and response matrix.
```

These models would fit in ~30 seconds with ergm 3.9. Getting these models to estimate would be a top priority, and then evaluating that the performance of the epidemic simulation (step 6 in the script) is consistent (i.e., unbiased and not slower) than the current approach is next.

**Comment by chad-klumb** [comment]
I triggered a rebuild on EpiModel@tergmLiteTerms and it now appears to work (Travis was having problems at the time I originally made the commits):

https://travis-ci.org/statnet/EpiModel/builds/535547522

**Comment by chad-klumb** [comment]
I remade the `EpiModel@tergmLiteTerms` branch so its changes sit on top of the present state of `master`

TravisCI passed and I think it's now amenable to merging

If you'd like me to do anything more for that branch please let me know

**Comment by smjenness** [comment]
I have merged `EpiModel@tergmLiteTerms` into master.

**Comment by smjenness** [comment]
I am closing out this issue because I have tested the new approach and it appears to work brilliantly. I am following up on other side comments as new issues here.
